receiver/awscontainerinsightreceiver/internal/k8sapiserver/k8sapiserver.go

```diff
@@ -16,54 +16,270 @@ package k8sapiserver
+				ci.Timestamp:      timestampNs,
+				ci.K8sNamespace:   namespace,
+				ci.Version:        "0",
+			}
+			if k.nodeName != "" {
+				attributes["NodeName"] = k.nodeName
+			}
+			md := ci.ConvertToOTLPMetrics(fields, attributes, k.logger)
+			result = append(result, md)
+		}
+	}
 	return result
 }
 
-func (k *K8sAPIServer) start() error {
-	//TODO: add implementation
+func (k *K8sAPIServer) init() error {
+	var ctx context.Context
+	ctx, k.cancel = context.WithCancel(context.Background())
+
+	k.nodeName = os.Getenv("HOST_NAME")
+	if k.nodeName == "" {
+		return errors.New("environment variable HOST_NAME is not set in k8s deployment config")
+	}
+
+	lockNamespace := os.Getenv("K8S_NAMESPACE")
+	if lockNamespace == "" {
+		return errors.New("environment variable K8S_NAMESPACE is not set in k8s deployment config")
+	}
+
+	configMapInterface := k.k8sClient.ClientSet.CoreV1().ConfigMaps(lockNamespace)
+	if configMap, err := configMapInterface.Get(ctx, lockName, metav1.GetOptions{}); configMap == nil || err != nil {
+		k.logger.Info(fmt.Sprintf("Cannot get the leader config map: %v, try to create the config map...", err))
+		configMap, err = configMapInterface.Create(ctx,
+			&v1.ConfigMap{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: lockNamespace,
+					Name:      lockName,
+				},
+			}, metav1.CreateOptions{})
+		k.logger.Info(fmt.Sprintf("configMap: %v, err: %v", configMap, err))
+	}
+
+	lock, err := resourcelock.New(
+		resourcelock.ConfigMapsResourceLock,
+		lockNamespace, lockName,
+		k.k8sClient.ClientSet.CoreV1(),
+		k.k8sClient.ClientSet.CoordinationV1(),
+		resourcelock.ResourceLockConfig{
+			Identity:      k.nodeName,
+			EventRecorder: k.createRecorder(lockName, lockNamespace),
+		})
+	if err != nil {
+		k.logger.Warn("Failed to create resource lock", zap.Error(err))
+		return err
+	}
+
+	go k.startLeaderElection(ctx, lock)
+
 	return nil
 }
 
-// Stop stops the k8sApiServer
-func (k *K8sAPIServer) Stop() {
+// Shutdown stops the k8sApiServer
+func (k *K8sAPIServer) Shutdown() {
 	if k.cancel != nil {
 		k.cancel()
 	}
 }
+
+func (k *K8sAPIServer) startLeaderElection(ctx context.Context, lock resourcelock.Interface) {
+
+	for {
+		leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
+			Lock: lock,
+			// IMPORTANT: you MUST ensure that any code you have that
+			// is protected by the lease must terminate **before**
+			// you call cancel. Otherwise, you could have a background
+			// loop still running and another process could
+			// get elected before your background loop finished, violating
+			// the stated goal of the lease.
+			LeaseDuration: 60 * time.Second,
+			RenewDeadline: 15 * time.Second,
+			RetryPeriod:   5 * time.Second,
+			Callbacks: leaderelection.LeaderCallbacks{
+				OnStartedLeading: func(ctx context.Context) {
+					k.logger.Info(fmt.Sprintf("k8sapiserver OnStartedLeading: %s", k.nodeName))
+					// we're notified when we start
+					k.leading = true
```

Comment:
Shouldn't the work to be performed happen inside this callback?  Also the `k.leading` value needs to be either guarded by a mutex or an atomic value as it might be changed concurrently.  If you really want this to just manage the `k.leading` flag so that `k.GetMetrics()` can short-circuit, then maybe something like this would work:

```go
				OnStartedLeading: func(ctx context.Context) {
					k.logger.Info(fmt.Sprintf("k8sapiserver OnStartedLeading: %s", k.nodeName))
					// we're notified when we start
					k.mu.Lock()
					k.leading = true
					k.mu.Unlock()

					if k.isLeadingC != nil {
						// this executes only in testing
						close(k.isLeadingC)
					}
					
					for {
						k.mu.Lock()
						leading := k.leading
						k.mu.Unlock()
						if !leading {
							k.logger.Info("no longer leading")
							return
						}
						select {
						case <-ctx.Done():
							k.logger.Info("ctx cancelled")
							return
						case <-time.After(time.Second):
						}
					}
				},
				OnStoppedLeading: func() {
					k.logger.Info(fmt.Sprintf("k8sapiserver OnStoppedLeading: %s", k.nodeName))
					// we can do cleanup here, or after the RunOrDie method returns
					k.mu.Lock()
					defer k.mu.Unlock()
					k.leading = false
					//node and pod are only used for cluster level metrics, endpoint is used for decorator too.
					k.k8sClient.Node.Shutdown()
					k.k8sClient.Pod.Shutdown()
				},
```

Reply:
Good suggestion! I will make the change.